### PR TITLE
Delete user

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

`Realm.App.deleteUser()` will soon be a requirement for Apple App Store: https://developer.apple.com/news/?id=i71db0mv

### ☑️ ToDos

* [x] 📝 Changelog entry
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* [x] 🚦 Tests
* ~[ ] 📱 Check the React Native/other sample apps work if necessary~
* ~[ ] 📝 Public documentation PR created or is not necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 🚦 Tests

### CODE REVIEW

File: docs/sync.js [hunk 1]
```diff
@@ -654,14 +664,6 @@ class ApiKeyAuth {
...
   fetchAll() {}
 
-  /**
-   * Deletes a user API key associated with the current user.
-   *
-   * @param {string} id - The id of the API key to delete.
-   * @returns {Promise<void>}
-   */
-  delete(id) {}
-
```
Comment (kraenhansen) [inline docs/sync.js:664]: Why was this removed?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,8 +1,7 @@
 x.x.x Release notes (yyyy-MM-dd)
 =============================================================
 ### Enhancements
-* None.
-
+* Added `Realm.User.delete()` to delete a sync user from a MongoDB Realm app. ([#4006](https://github.com/realm/realm-js/issues/4006))
```
Comment (kraenhansen) [inline CHANGELOG.md:4]:
```suggestion
* Added `Realm.App#deleteUser(user)` to delete a sync user from a MongoDB Realm app. ([#4006](https://github.com/realm/realm-js/issues/4006))
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -46,8 +45,13 @@ Please note the following API changes from the `10.12.0-beta.1` release of Flexi
 ### Fixed
 * Opening a Realm with a schema that has an orphaned embedded object type performed an extra empty write transaction. ([realm/realm-core#5115](https://github.com/realm/realm-core/pull/5115), since v10.5.0)
 * Schema validation was missing for embedded objects in sets, resulting in an unhelpful error being thrown if the user attempted to define one. ([realm/realm-core#5115](https://github.com/realm/realm-core/pull/5115))
+<<<<<<< HEAD
 * Fixed a crash when closing an Electron app with a custom sync error handler ([#4150](https://github.com/realm/realm-js/issues/4150)
 * Adding data from authentication providers, to be included in the `User#profile` object. ([#3268](https://github.com/realm/realm-js/issues/3268), since v10.0.0)
+=======
+* Improved error messaging when opening a Realm with incompatible histories. ([realm/realm-core#5161](https://github.com/realm/realm-core/pull/5161), since v6.0.0)
+* UserIdentity metadata table grows indefinitely. ([realm/realm-core#5152](https://github.com/realm/realm-core/issues/5152), since v10.0.0)
+>>>>>>> dc91ffda (wip)
```
Comment (kraenhansen) [inline CHANGELOG.md:54]: I guess this needs to be resolved before merging.

File: packages/realm-web/CHANGELOG.md [hunk 4]
```diff
@@ -5,7 +5,7 @@
 * None
 
 ### Enhancements
-* None
+* Added `Realm.User.delete()` to delete a sync user from a MongoDB Realm app. ([#4006](https://github.com/realm/realm-js/issues/4006))
```
Comment (kraenhansen) [inline packages/realm-web/CHANGELOG.md:8]:
```suggestion
* Added `Realm.App#deleteUser(user)` to delete a sync user from a MongoDB Realm app. ([#4006](https://github.com/realm/realm-js/issues/4006))
```

File: docs/sync.js (hunk 1 above)
Comment (kneth) [reply docs/sync.js:664]: Clearly a mistake. Well spotted!

File: packages/realm-web/src/App.ts [hunk 5]
```diff
@@ -175,7 +175,7 @@ export class App<
       this.storage.clear();
       // A failed hydration shouldn't throw and break the app experience
       // Since this is "just" persisted state that unfortunately got corrupted or partially lost
-      console.warn("Realm app hydration failed:", err.message);
+      console.warn("Realm app hydration failed:", (err as Error).message);
```
Comment (kraenhansen) [inline packages/realm-web/src/App.ts:178]:
```suggestion
      console.warn("Realm app hydration failed:", err instanceof Error ? err.message : err);
```

File: src/js_user.hpp [hunk 6]
```diff
@@ -30,6 +30,7 @@
 #include <realm/object-store/sync/sync_session.hpp>
 #include <realm/object-store/sync/sync_user.hpp>
 #include <realm/object-store/sync/app.hpp>
+#include "js_types.hpp"
```
Comment (kraenhansen) [inline src/js_user.hpp:33]: Is this still needed?

File: tests/js/user-tests.js [hunk 7]
```diff
@@ -548,4 +548,52 @@ module.exports = {
...
+    const validPassword = "test1234567890";
+    await app.emailPasswordAuth.registerUser({ email: validEmail, password: validPassword });
+    let user1 = await app.logIn(Realm.Credentials.emailPassword(validEmail, validPassword));
+
+    TestCase.assertTrue(user1.isLoggedIn);
+
+    app.removeUser(user1);
+    TestCase.assertFalse(user1.isLoggedIn);
+
+    // the user still exists
```
Comment (kraenhansen) [inline tests/js/user-tests.js:567]:
```suggestion
    // Expect that the user still exists on the server
```

**Review by kraenhansen — APPROVED** [review]
I have only a couple of minor comments.

File: CHANGELOG.md [hunk 8]
```diff
@@ -57,7 +54,7 @@ Please note the following API changes from the `10.12.0-beta.1` release of Flexi
 * File format: generates Realms with format v22 (reads and upgrades file format v5 or later for non-synced Realm, upgrades file format v10 or later for synced Realms).
 
 ### Internal
-* Upgraded Realm Core from v11.7.0 to v11.8.0. ([#4207](https://github.com/realm/realm-js/issues/4207))
+* Upgraded Realm Core from v11.7.0 to v11.9.0.
```
Comment (fronck) [inline CHANGELOG.md:57]:
```suggestion
* Upgraded Realm Core from v11.7.0 to v11.8.0.
```
This seems like a merge error..?

File: src/js_app.hpp [hunk 9]
```diff
@@ -291,6 +293,30 @@ void AppClass<T>::remove_user(ContextType ctx, ObjectType this_object, Arguments
     app->remove_user(*user, Function::wrap_void_callback(ctx, this_object, callback));
 }
 
+/**
+ * @brief Delete permanently the user server-side. Moreover, the user will be logged out
+ *        locally, and the current user will be changed. The function is asynchronously
```
Comment (fronck) [inline src/js_app.hpp:298]:
```suggestion
 *        locally, and the current user will be changed. The function is asynchronous
```

**Review by fronck — APPROVED** [review]
Please see the minor comments.

File: docs/sync.js [hunk 10]
```diff
@@ -677,6 +678,14 @@ class ApiKeyAuth {
...
    */
   disable(id) {}
+
+  /**
+   * Deletes a user API key associated with the current user.
+   *
+   * @param {string} id - The id of the API key to delete.
+   * @returns {Promise<void>}
+   */
+  delete(id) {}
```
Comment (takameyer) [inline docs/sync.js:688]: I would like this method to be called `deleteApiKey`, but that would be a breaking change.

File: packages/realm-web-integration-tests/harness/index.ts [hunk 11]
```diff
@@ -40,6 +40,9 @@ const EXPECTED_ISSUES = [
   /was set with `SameSite=None` but without `Secure`/,
   // User / "refresh invalid access tokens" is posting an invalid token
   "Failed to load resource: the server responded with a status of 401 (Unauthorized)",
+  // User / "can be deleted" is deleting an invalid token and attempting a failing login
+  "Failed to load resource: the server responded with a status of 401 (Unauthorized)",
+  "Failed to load resource: the server responded with a status of 401 (Unauthorized)",
```
Comment (takameyer) [inline packages/realm-web-integration-tests/harness/index.ts:45]: i would like some explanation about what this data structure is doing and why there are multiples of the string in it.

Comment (takameyer) [reply packages/realm-web-integration-tests/harness/index.ts:45]: It seems like the `checkIssues` function could be refactored using the array methods `some` and `includes` and that would probably remove the need to have the same "expected issues" listed in this array.

File: packages/realm-web/src/App.ts [hunk 12]
```diff
@@ -247,6 +247,17 @@ export class App<
     this.storage.removeUserId(user.id);
   }
 
+  /**
+   * @inheritdoc
+   */
+  public async deleteUser(user: User<FunctionsFactoryType, CustomDataType>): Promise<void> {
+    await this.fetcher.fetchJSON({
```
Comment (takameyer) [inline packages/realm-web/src/App.ts:254]: Do we need some sort of error handling here?

File: packages/realm-web/src/tests/App.test.ts [hunk 13]
```diff
@@ -355,6 +355,54 @@ describe("App", () => {
...
+    const app = new App({
+      id: "my-mocked-app",
+      storage,
+      transport,
+      baseUrl: "http://localhost:1234",
+    });
+    const credentials = Credentials.emailPassword("gilfoyle@testing.mongodb.com", "v3ry-s3cret");
+    const user = await app.logIn(credentials);
+
+    // Expect logging in returns a user
```
Comment (takameyer) [inline packages/realm-web/src/tests/App.test.ts:397]:
```suggestion
    // Expect login returns a user
```

File: src/js_app.hpp [hunk 14]
```diff
@@ -291,6 +293,30 @@ void AppClass<T>::remove_user(ContextType ctx, ObjectType this_object, Arguments
     app->remove_user(*user, Function::wrap_void_callback(ctx, this_object, callback));
 }
 
+/**
+ * @brief Delete permanently the user server-side. Moreover, the user will be logged out
```
Comment (takameyer) [inline src/js_app.hpp:297]:
```suggestion
 * @brief Permanently delete the user server-side. Moreover, the user will be logged out
```

**Review by takameyer — APPROVED** [review]
Looks good.  Just a few comments/suggestions

File: docs/sync.js [hunk 15]
```diff
@@ -150,6 +150,15 @@ class App {
    */
   removeUser(user) {}
 
+  /**
+   * Deletes the user.
+   * NOTE: This irrecoverably deletes the user from device as well as the server!
```
Comment (tomduncalf) [inline docs/sync.js:155]:
```suggestion
   * NOTE: This irrecoverably deletes the user from the device as well as the server!
```

File: types/app.d.ts [hunk 16]
```diff
@@ -331,6 +331,14 @@ declare namespace Realm {
      * @returns A promise that resolves once the user has been logged out and removed from the app.
      */
     removeUser(user: User<FunctionsFactoryType, CustomDataType>): Promise<void>;
+
+    /**
+     * Delete the user.
+     * NOTE: This irrecoverably deletes the user from device as well as the server!
```
Comment (tomduncalf) [inline types/app.d.ts:337]:
```suggestion
     * NOTE: This irrecoverably deletes the user from the device as well as the server!
```

File: docs/sync.js (hunk 10 above)
Comment (kneth) [reply docs/sync.js:688]: Please create an issue :smile:

File: packages/realm-web/src/App.ts (hunk 12 above)
Comment (kneth) [reply packages/realm-web/src/App.ts:254]: @kraenhansen What do you think?

File: docs/sync.js (hunk 10 above)
Comment (takameyer) [reply docs/sync.js:688]: Done https://github.com/realm/realm-js/issues/4322 :)

File: packages/realm-web-integration-tests/harness/index.ts (hunk 11 above)
Comment (kraenhansen) [reply packages/realm-web-integration-tests/harness/index.ts:45]: Does the documentation comment above clear things up?

> A list of expected errors and warnings logged to the console

It's there to ensure our SDK are not triggering unexpected errors or warnings in the console.

The reason the same message is in the array more than once, is because we don't want to allow an unlimited number of messages (even if they have the same message), as they might be caused by some other (potentially new) regression.

If we had to refactor this, we should probably move these messages into the tests somehow. Make them a property on the tests or something:

```typescript
it("prints an error", function() {
  this.expectedErrors([ "the error" ]);
  console.error("the error");
});
```

File: src/js_user.hpp (hunk 6 above)
Comment (kneth) [reply src/js_user.hpp:33]: It looks like I can build without it --> will remove the line.

File: packages/realm-web/src/App.ts (hunk 12 above)
Comment (takameyer) [reply packages/realm-web/src/App.ts:254]: I mean, if `fetchJSON` throws, then it would probably be clear in the stack trace that there was an issue with `deleteUser`.  Just want to make sure this doesn't fail silently on a network issue

Comment (kraenhansen) [reply packages/realm-web/src/App.ts:254]: Nothing fails silently here. We await the result of `fetchJSON` so any rejection will be propagated as expected.

File: packages/realm-web-integration-tests/harness/index.ts (hunk 11 above)
Comment (takameyer) [reply packages/realm-web-integration-tests/harness/index.ts:45]: So I've looked into this deeper and believe that it would be much more useful to refactor the tests with the proposed syntax from Kraen:

> If we had to refactor this, we should probably move these messages into the tests somehow. Make them a property on the tests or something:
```typescript
it("prints an error", function() {
  this.expectedErrors([ "the error" ]);
  console.error("the error");
});
```
In my quick study, I found that removing one of the elements from the array caused the test harness to fail, without any hint of why, other than that some log was received at some point that it wasn't expecting.  Putting this into the test itself would help make this more maintainable in the future.

Comment (kneth) [reply packages/realm-web-integration-tests/harness/index.ts:45]: I have created https://github.com/realm/realm-js/issues/4323 to track the work.

File: packages/realm-web/src/App.ts (hunk 12 above)
Comment (kneth) [reply packages/realm-web/src/App.ts:254]: I will leave it as it is.

File: packages/realm-web-integration-tests/harness/index.ts (hunk 11 above)
Comment (takameyer) [reply packages/realm-web-integration-tests/harness/index.ts:45]: It seems the main blocker in fixing this is that it's unclear how to pass the log messages from the headless chrome into the test suites.  Perhaps one would have to setup some sort of channel to transmit the log messages from the log event to the mocha client.  In any case, not so trivial
